platform_design/lib/main.dart

```diff
@@ -0,0 +1,205 @@
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/scheduler.dart';
+
+import 'tab_1.dart';
+import 'tab_2.dart';
+import 'tab_3.dart';
+import 'tab_4.dart';
+import 'widgets.dart';
+
+void main() => runApp(MyApp());
+
+class MyApp extends StatelessWidget {
+  final platformOverride = ValueNotifier<TargetPlatform>(defaultTargetPlatform);
+
+  @override
+  Widget build(BuildContext context) {
+    // Change this value to better see animations.
+    timeDilation = 1;
+    // Either Material or Cupertino widgets work in either Material or Cupertino
+    // Apps.
+    return ValueListenableBuilder(
+      valueListenable: platformOverride,
+      child: PlatformAdaptingHomePage(platformOverride: platformOverride),
+      builder: (BuildContext context, TargetPlatform platform, Widget child) {
+        return MaterialApp(
+          title: 'Adaptive Music App',
+          theme: ThemeData(
+            // Use the green theme for Material widgets.
+            primarySwatch: Colors.green,
+            platform: platform,
+          ),
+          builder: (BuildContext context, Widget child) {
+            return CupertinoTheme(
+              // Instead of letting Cupertino widgets auto-adapt to the Material
+              // theme (which is green), we're going to use a different theme
+              // for Cupertino (which is blue by default).
+              data: CupertinoThemeData(),
+              child: Material(child: child),
+            );
+          },
+          home: child,
+        );
+      },
+    );
+  }
+}
+
+// The widget which shows a different type of scaffold depending on the platform.
+//
+// This file has the most amount of non-sharable code since it behaves the most
+// differently between the platforms.
+class PlatformAdaptingHomePage extends StatefulWidget {
+  PlatformAdaptingHomePage({ Key key, this.platformOverride }) : super(key: key);
+
+  final ValueNotifier<TargetPlatform> platformOverride;
+
+  @override
+  _PlatformAdaptingHomePageState createState() => _PlatformAdaptingHomePageState();
+}
+
+class _PlatformAdaptingHomePageState extends State<PlatformAdaptingHomePage> {
+  // We're keeping global keys because tabs 1 and 2 own a bunch of data.
+  // We'd like to keep them as we change platforms and reparent those tabs
+  // into different scaffolds.
+  final tab1Key = GlobalKey();
+  final tab2Key = GlobalKey();
+
+  // In Material, we're using the hamburger menu paradigm. We're flatly listing
+  // all possible tabs and injecting this builder into tab 1 which is building
+  // the scaffold around the drawer.
+  Widget _buildAndroidDrawer(BuildContext context) {
```

Comment:
It should. It would be more idiomatic. I was just too lazy to try to figure out whether that new widget declaration should come before PlatformAdaptingHomePage or after to make the readability flow better.

Reply:
I left the comment here for symmetry but moved the widget implementation below.